src/libxml/node_manip.cxx

```diff
@@ -45,6 +45,17 @@
 namespace
 {
 
+// Helper: sets the given namespace for this node and all of its children,
+// recursively, but without overriding any existing namespaces.
+void set_default_ns_recursively(xmlNodePtr node, xmlNsPtr ns)
+{
+    if ( !node->ns )
+        node->ns = ns;
+
```

Comment:
Shouldn't the recursion stop here if `node->ns` is set?

The argument for implicitly setting the NS is that the NS should be the same as the parent node's if not set — but if you descend under a node that *does* have a NS set, it would be surprising to set its children's NS implicitly to the grand-parent ones:

```
<NS_A:a>
  <foo></foo>  <!— implicitly set to NS_A —>
  <NS_SUB:child>
    <bar> <!— implicitly set to NS_A, but I would expect NS_SUB —>
  </NS_SUB:child>
</NS_A:a>
```

Of course, as long as the tree is constructed with xmlwrapp, `<bar>` in the above example will have correct NS assigned already (from when it was inserted into `<child>`) — but that's another reason to terminate recursion to do less work, isn't it?

Reply:
Yes, you're absolutely right, of course, thanks! PR updated.